atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheImplTest.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.api.watch;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.palantir.atlasdb.transaction.api.TransactionLockWatchFailedException;
+import com.palantir.lock.AtlasRowLockDescriptor;
+import com.palantir.lock.LockDescriptor;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.watch.CacheStatus;
+import com.palantir.lock.watch.CommitUpdate;
+import com.palantir.lock.watch.IdentifiedVersion;
+import com.palantir.lock.watch.ImmutableTransactionUpdate;
+import com.palantir.lock.watch.LockEvent;
+import com.palantir.lock.watch.LockWatchCreatedEvent;
+import com.palantir.lock.watch.LockWatchEvent;
+import com.palantir.lock.watch.LockWatchReferences;
+import com.palantir.lock.watch.LockWatchStateUpdate;
+import com.palantir.lock.watch.TransactionUpdate;
+
+@RunWith(MockitoJUnitRunner.class)
+public final class LockWatchEventCacheImplTest {
```

Comment:
I will put one comment for LockWatchEventCacheImplTest and LockWatchEventLogImplTest (and kinda LockWattchEventCacheIntegrationTest):

I feel that the tests in LockWatchEventCacheImplTest and LockWatchEventLogImplTest, where you mock all the datastructures underneath them, are quite fragile in general. What I kinda had in mind is:

1. LockWatchEventCacheImpl has a package private method called #getSnapshot which returns a Snapshot of all the internal datastructures (so it contains log of events, all the open transactions and their state, and current snapshot). That thing is Json serializable
2. You only have 1 big integration test, that tests a bunch of scenarios where you poke the LockWatchEventCacheImpl, but after each step take a snapshot and compare it to a golden file (and do as many steps as you want).
3. The review is just a matter of: read a test case in LockWAtchEventCacheImplIntegrationTest, understand what it should do to the internal datastructures, and go diff each step and see that the right thing happened.

This is a bit of extra code that will only be used for tests, however I would argue that the implementation of this cache is the critical piece to all of lock watches (this and the timelock code).

Reply:
I think this does describe a useful test, although it does depend on the creation of the "golden file" being correct: given that this is the first implementation, I feel like it is hard to convince oneself that "this golden file is actually 100% super correct", other than by manual review. Still, I absolutely agree about the tests being pretty fragile, and I think this could be pretty good. I'll experiment and see what it looks like.